pybamm/expression_tree/array.py

```diff
@@ -1,12 +1,17 @@
 #
 # NumpyArray class
 #
+from __future__ import annotations
 import numpy as np
 from scipy.sparse import csr_matrix, issparse
+from typing import TYPE_CHECKING
 
 import pybamm
 from pybamm.util import have_optional_dependency
 
+if TYPE_CHECKING:  # pragma: no cover
+    import sympy
```

Comment:
Why do we need to enable type checking with a global variable?

Reply:
Related to above - this is a statement used by mypy. In this case sympy is used in the file only as a type hint, not during runtime. `TYPE_CHECKING` is a constant assumed to be True by static type checkers, but is false at runtime; so sympy will only be imported if a type checker like mypy is being run. https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Reply:
Personally I would prefer to just have sympy always imported. If it is used in the type annotations, then why not just always have it there?

Reply:
That would make sympy a required dependency. Sympy is heavy and I would avoid making it a required dependency (only a subset of features in PyBaMM require sympy)

Reply:
I would say that if expression tree is using sympy types, then it is really a dependency. I have not downloaded this branch locally to double check, but I would guess that my IDE would complain if I did not have sympy installed too.

As we add type checking it is going to expose more things that should probably be real dependencies rather than optional ones.

Reply:
sympy is only used in the expression tree for `to_equation()` to use with LaTeX I believe, rather than the solvers, hence the optional dependency.

Reply:
Yeah my point was more that we might have to rethink our optional dependencies if we are using static types. Expression tree depends on sympy if it is using sympy types, but by putting this check here we are pretending that it doesn't depend on sympy.

It is too much for this task, but if we want a typed library and not have all the dependencies for types then we might need to think about how to better separate our subcomponents. Maybe things like the latexify need to go into a separate module if we want it's dependencies to be truly optional